Add apimachinery package for parsing manifests

errordeveloper opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

    
When using Kubernetes client it's not very easy
to read a manifest with just any set of resource
from file and parse it in the same way kubectl
would do. This code was previously used by eksctl
project and could be useful more widely.
    
This commit imports code from [eksctl repository][1].
    
The author of this commit is also the original author
of the given code. It was originally under APL2, hence
there should be no licensing issues either.
    
[1]: https://github.com/weaveworks/eksctl/blob/91b4eddc988276c8a3c857321e1347bd7405d954/pkg/kubernetes/manifests.go

**Special notes for your reviewer**:

The author of this commit is also the original author
of the given code. It was originally under APL2, hence
there should be no licensing issues either.

**Does this PR introduce a user-facing change?**:

```release-note
Add apimachinery package for parsing manifests
```

errordeveloper committed:
Add apimachinery package for parsing manifests

When using Kubernetes client it's not very easy
to read a manifest with just any set of resource
from file and parse it in the same way kubectl
would do. This code was previously used by eksctl
project and could be useful more widely.

This commit imports code from eksctl repository [1].

The author of this commit is also the original author
of the given code. It was originally under APL2, hence
there should be no licensing issues either.

[1]: https://github.com/weaveworks/eksctl/blob/91b4eddc988276c8a3c857321e1347bd7405d954/pkg/kubernetes/manifests.go

errordeveloper commented:
/cc @pwittrock @soltysh @justinsb

errordeveloper commented:
/test pull-kubernetes-e2e-kind

errordeveloper commented:
/test pull-kubernetes-e2e-gce

luxas commented:
Interesting, thanks @errordeveloper!

The past couple of weeks myself and @twelho have been working on a even more high-level serializer (or whatever we would call it), which does most of this (and a fair amount more). It's not a 1:1 mapping, but I think that we need this kind of thing in k8s core repos somewhere.

https://pkg.go.dev/github.com/weaveworks/libgitops/pkg/serializer?tab=doc has some description, and if you create a decoder `WithListElementsDecoding(true)` it will automatically decode v1.Lists into objects. It's a different level, as it more operates on `runtime.Objects` (and mostly not `runtime.RawExtensions`), and uses a given scheme instead of client-go, but it might be interesting to look at this in tandem. I was planning to donate this code to upstream anyways.

pwittrock commented:
@errordeveloper Thanks for this PR!  I think this needs to wait until after kubectl is out of the mono-repo due to circular deps.  IIRC eta is 1-2 releases.

cc @soltysh

pwittrock commented:
May be the apimachinery libraries are stable enough now that this is ok?  Probably worth revisiting.

pwittrock commented:
cc @lavalamp

lavalamp commented:
The referenced apimachinery libraries don't make promises we won't break you, but in practice we've not broken users of them in quite a while, at least to my knowledge.

If this code is just about turning a yaml `---` separated doc into a list of runtime.Objects, I'd actually consider that for inclusion somewhere in an apimachinery library. I won't have time to look at such a thing immediately, unfortunately.

errordeveloper commented:
Thanks folks for chummin#g in!

I am not super clear on what would be the best home for this, and overall spirit of this PR was that this is indeed needed somewhere.

When I wrote this code in eksctl, I needed to read just any objects from a file that kubectl would accept. More generally, I just needed to do some of the things that kubectl does. It is hard to setup all the right bits from scratch, and I think it should be easier.

Aside from this code, I managed to use cli-runtime helpers, and was considering to contribute those also, but perhaps cli-runtime could use some reworking to become more user-friendly.

The reasons I got back to this code is because I needed to do something similar again, but in an operator context this time. With a few tips from @JoelSpeed I realised that `UnsturcuredList.UnmarshallJSON` does the trick (albeit only for JSON lists(?)), and controller-runtime client handles any objects just fine. It does seem like controller-runtime/cli-runtime is largely an artificial split, the should be a more general-purpose client that is easy enough to use.

One of the challenges user has with this type of problems is the whole galore of types - `Unstructured`, `RawExtension`, and what else there is... What is most unclear is how to construct these and `v1.List`.

I think code in this PR could be worse something, but largely my money is on making it easier (or at least clearer) how does one write a Go program that handles some of the things that kubectl does as far as parsing multi-object manifests is concerned, and handing objects to the API in one manner or another.
What I mean by clearer/easier is this - reading kubeconfigs is not easy, but it's clear how does one do it normally. It looks like controller-runtime has a few clear ways already.

Also, not in a hurry to move this in, just wanted to get the ball rolling. My present use-case (CR -> CUE -> JSON -> more CRs -> API) appears to be fullfiled by controller-runtime.

lavalamp commented on staging/src/k8s.io/cli-runtime/pkg/manifest/manifest.go:
```diff
@@ -0,0 +1,86 @@
+
+// NewRawExtensions decodes the provided manifest's bytes into "raw extension"
+// Kubernetes objects. These can then be passed to NewRawResource.
+func NewRawExtensions(manifest []byte) ([]runtime.RawExtension, error) {
```
I think we mostly use Unstructured rather than RawExtension when the context isn't embedding an object.

errordeveloper replied:
Thanks, I'll try to change this, to be honest these types are not very self-descriptive...

lavalamp commented on staging/src/k8s.io/cli-runtime/pkg/manifest/manifest.go:
```diff
@@ -0,0 +1,86 @@
+	if component.Object != nil {
+		gvk := component.Object.GetObjectKind().GroupVersionKind()
+		if strings.HasSuffix(gvk.Kind, "List") {
+			// must use *corev1.List, as it cannot be converted to *metav1.List
```
Well, there should be ways. I think you'd have to lose the corev1 import before this code could go in apimachinery.

errordeveloper replied:
Makes sense, I'll see how this could be done, it's possible that `UnstructureList` is what I want here.

lavalamp commented:
I don't have a more specific suggestion at the moment than somewhere under `k8s.io/apimachinery/pkg/`. I left a couple comments.

errordeveloper commented:
/assign @lavalamp

errordeveloper commented:
@lavalamp I managed to simply this quite a bit also, looking much better now :)

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,72 @@
+)
+
+// NewUnstructuredList decodes YAML or JSON data into a flattened list of Kubernetes resources,
+// it will attempt parsing all manner of manifests, such as YAML multi docs and any list types
```
I'm a little skeptical that many people will want to flatten multidocs with mixed lists and loose objects. I recommend making this two steps, e.g. read the multidoc into a list, then optionally call a "flatten" function to expand embedded lists. But I'd like to see a use case for that latter step before putting it in a general library, since it's easy to add stuff here and impossible to remove it :)

errordeveloper replied:
Here's a very simple example:
```
---
apiVersion: v1
kind: List
items:
- apiVersion: v1
   kind: Namespace
   metadata:
       name: my-namespace
```
Having parsed this to `unstructured.UnstructuredList` you have no way of reasoning about the depth and it's too easy to forget that it can be deeply nested, or even no too deeply as per above example.

I agree that it doesn't have a separate explicit step or a a boolean function argument would be a more sensible way to implement this.

The main challenge is really - how does a user read all objects from a manifest and iterate over them to pass them to a client, and they very little about internal representation, but they are familiar with `k8s.io/api`, and they might presume that there's some type that holds abstract objects (like `unstructured.Unstructured`).

errordeveloper replied:
@lavalamp I've changed it to not call `AppendFlattened`

lavalamp replied:
Comment needs an update now :)

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest_test.go:
```diff
@@ -0,0 +1,67 @@
+	}{
+		{
+			path:        "testdata/misc-sample-nested-list-1.json",
+			expectedLen: 6,
```
The distance between the data and this expectation makes this test hard to read. I don't have an exact suggestion, maybe add the expectation to the file name or somehow embed the expectation in the file?

errordeveloper replied:
I've not changes this yet, let's settle the API first :)

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+
+// AppendFlattened will append newItem to list; making sure that raw newItem is decoded
+// and flattened with another list; it drops empty lists
+func AppendFlattened(list *unstructured.UnstructuredList, newItem *unstructured.Unstructured) error {
```
Can we make this private? I suspect most people only need to call Flatten below?

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+	return nil
+}
+
+// Flatten will ensure given list is fully flattened
```
Can we expand this comment and cover why someone might want to do this, and the fact that it loses the nesting structure?

luxas commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+	// IsList() checks if object has 'items', but it will ignore
+	// something like `{"apiVersion": "v1", "kind": "List"}`, which
+	// is still a list and should be treated as such
+	if newItem.IsList() || strings.HasSuffix(gvk.Kind, "List") {
```
I would check both for `apiVersion: v1` and `kind: List`, and make this line a private separate `isList` function for easier unit-testing. Also, `"List"` and `"v1"` I'd make constants.

luxas commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+	if newItem.Object == nil {
+		return nil
+	}
+	gvk := newItem.GetObjectKind().GroupVersionKind()
```
Catch the case of the gvk being empty and error?

luxas commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+limitations under the License.
+*/
+
+package kubernetes
```
this package is called manifest, right?

errordeveloper replied:
Well, spotted, thanks!

luxas commented on staging/src/k8s.io/apimachinery/pkg/util/manifest/manifest.go:
```diff
@@ -0,0 +1,81 @@
+
+// NewUnstructuredList decodes YAML or JSON data into a flattened list of Kubernetes resources,
+// it will attempt parsing all manner of manifests, such as YAML multi docs and any list types
+func NewUnstructuredList(data []byte) (*unstructured.UnstructuredList, error) {
```
Shall we name this `(Parse|Unmarshal|Decode)UnstructuredList`? I guess `Parse...` would make the most sense.

luxas replied:
I guess it's better also to accept an `io.ReadCloser` instead of bytes here, to allow multiple kinds of sources?

errordeveloper replied:
If taking a reader is important, I'd insist on two functions (one with `[]bytes` and another with a reader).

luxas reviewed (commented):
With only two functions (`ParseUnstructuredList` and `Flatten`), this looks really clean :+1:.

Now the question becomes, should this be part of the `unstructured` package instead, as this is specifically operating solely on those objects?

errordeveloper commented:
/remove-lifecycle rotten

On Fri, 29 Jan 2021, 5:17 pm fejta-bot, <notifications@github.com> wrote:

> Stale issues rot after 30d of inactivity.
> Mark the issue as fresh with /remove-lifecycle rotten.
> Rotten issues close after an additional 30d of inactivity.
>
> If this issue is safe to close now please do so with /close.
>
> Send feedback to sig-testing, kubernetes/test-infra and/or fejta
> <https://github.com/fejta>.
> /lifecycle rotten
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/92885#issuecomment-769935302>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5MS7GNX5X5RDWORIWR5TS4LUSRANCNFSM4OTPPNOA>
> .
>

lavalamp commented:
I think this is waiting on another push addressing comments, is that correct?

errordeveloper commented:
Yes, just didn't get around to it yet, but I am hopeful!

On Fri, 29 Jan 2021, 10:18 pm Daniel Smith, <notifications@github.com>
wrote:

> I think this is waiting on another push addressing comments, is that
> correct?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/92885#issuecomment-770080624>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAB5MS74XWMALMAY336O6ULS4MX5BANCNFSM4OTPPNOA>
> .
>

luxas commented:
I've been thinking about and exploring this area in quite some depth, and have some k8s project spanning ideas on how to deal with JSON and YAML more consistently, and support all the features the ecosystem needs, depends on and expects. I'm planning to write a KEP about this, and regularly join the SIG API Machinery meetings to get feedback on the direction.

So hopefully, after a document outlining unifications that could be made across most of our ecosystem, we can get going with this type of change again. Let me know if you're interested in working with me on this task. I'll let you know more during the next SIG API Machinery meeting on the 22nd :smile:. Meanwhile, if you have questions, please do ask e.g. on Slack!

errordeveloper commented:
@luxas I've been meaning to pick this up again, but didn't have the cycles, as I'm not actively working on something that depends on this at present (I used to have a related project, and don't do much with it any more). Please do keep up to date, I might find some cycles in the future.

luxas commented:
No problem. I can cc you on the proposal later when written, it'd be great to make sure that what you needed fits within the scope of that KEP :+1:
